bluej/src/main/java/bluej/parser/context/CompilationUnitContextLoader.java

```diff
@@ -0,0 +1,378 @@
+     *
+     * @param qualifiedName The fully qualified class name
+     * @param info The ClassInfo from the parser
+     *
+     * @return A new CompilationUnitContext with the data from ClassInfo
+     */
+    @NotNull public CompilationUnitContext updateContextFromClassInfo(@NotNull String qualifiedName, @NotNull Path projectDir, @NotNull ClassInfo info) {
+        // Extract ClassInfo data first (handles FX thread requirements)
+        ClassInfoData data = extractClassInfoData(info);
+
+        CompilationUnitContext context = CompilationUnitContext.readOnly(data.className);
+
+        context.setComments(PropertyContextFormat.fromProperties(data.comments));
+
+        Path contextFilePath = projectDir.resolve(constructContextFilePath(qualifiedName));
+
+        try {
+            File contextFile = contextFilePath.toFile();
+
+            PropertyContextFormat.writeToFile(context, contextFile);
+        } catch (IOException ioe) {
+            throw new UncheckedIOException("Could not write context for " + contextFilePath, ioe);
+        }
+
+        // Invalidate cache entry so subsequent lookups reload the updated data
+        evictFromCache(qualifiedName);
+
+        return context;
+    }
+
+
+    /**
+     * Invalidates the cached context for a class.
+     * This method searches through all registered package roots to find
+     * and evict the corresponding cache entry.
+     *
+     * @param qualifiedName The fully qualified class name
+     * @return true if an entry was removed from cache, false otherwise
+     */
+    public boolean evictFromCache(@NotNull String qualifiedName) {
+        boolean evicted = false;
+
+        if (cachingEnabled) {
+            cache.remove(qualifiedName);
+        }
+
+        return evicted;
+    }
+
+
+    /**
+     * Clears the cache if caching is enabled.
+     * No-op if caching is disabled.
+     *
+     * <p>This method is thread-safe and can be called while other threads
+     * are accessing the cache.
+     */
+    public void clearCache() {
+        if (cache != null) {
+            cache.clear();
+        }
+    }
+
+
+    /**
+     * Closes this loader and releases any cached resources.
+     * This method clears the cache to free memory.
+     *
+     * <p>After closing, the loader can still be used but will need to
+     * reload any previously cached contexts.
+     *
+     * <p>This method is idempotent and thread-safe.
+     */
+    @Override
+    public void close() {
+        clearCache();
+    }
+
+
+    /**
+     * Helper method to safely extract ClassInfo data on the FX thread.
+     * This method ensures that ClassInfo.getName() and ClassInfo.getComments()
+     * are called on the JavaFX platform thread as required.
+     *
+     * @param info The ClassInfo to access
+     * @return A ClassInfoData record containing the extracted data
+     */
+    static ClassInfoData extractClassInfoData(@NotNull ClassInfo info) {
```

Comment:
This was kind of annoying because of the thread checker. It complained I was not doing this on the JavaFX thread, but then when I added the future it turns out I acutally was on the thread and it deadlocked xD

In the end I annotated things I've needed with `Tag.Any` and changed this code so it works both on and off the JavaFX thread, but I'm not sure if that's the best solution. Any suggestions?